main: safely check stop_signal in-between starting services

bhalevy opened the pull request:
To simplify aborting scylla while starting the services,
add a _ready state to stop_signal, so that until
main is ready to be stopped by the abort_source,
just register that the signal is caught, and
let a check() method poll that and request abort
and throw respective exception only then, in controlled
points that are in-between starting of services
after the service started successfully and a deferred
stop action was installed.

This patch prevents gate_closed_exception to escape handling
when start-up is aborted early with the stop signal,
causing https://github.com/scylladb/scylladb/issues/23153
The regression is apparently due to a25c3eaa1caa9826abba750dc004566bdda89eb6

Fixes https://github.com/scylladb/scylladb/issues/23153

* Requires backport to 2025.1 due to a25c3eaa1caa9826abba750dc004566bdda89eb6

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15662/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3712/)
:x: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6827/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14851/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15662/)
#### Build Details:
- Duration: 5 hr 2 min
- Builder: spider5.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15662/) ❌ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3712/) ❌ - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6827/) ✅ - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14851/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15662/)

The dtest needs to ignore yet another error message: see https://github.com/scylladb/scylla-dtest/pull/5590

bhalevy commented:
[ce8d3fe](https://github.com/scylladb/scylladb/commit/ce8d3fe7345b51e210e3af86c728f22e0d0183e4):
- minor update to task manager start message

```diff
diff --git a/main.cc b/main.cc
index a8bad807ae..098fcd32c1 100644
--- a/main.cc
+++ b/main.cc
@@ -1141,7 +1141,7 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             }
             view_hints_dir_initializer.ensure_created_and_verified().get();
 
-            safe_start(stop_signal, "task manager");
+            safe_start(stop_signal, "starting task manager");
             auto get_tm_cfg = sharded_parameter([&] {
                 return tasks::task_manager::config {
                     .task_ttl = cfg->task_ttl_seconds,
```

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15680/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14871/)
:x: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6848/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3732/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15680/)
#### Build Details:
- Duration: 4 hr 56 min
- Builder: spider6.cloudius-systems.com

denesb commented:
Backports need an issue + fixes.

denesb commented on main.cc:
```diff
@@ -537,6 +553,11 @@ void print_starting_message(int ac, char** av, const bpo::parsed_options& opts)
     fmt::print("parsed command line options: {}\n", format_parsed_options(opts.options));
 }
 
+static auto safe_start(stop_signal& stop, sstring what, bool ready = false) {
```
this is not starting anything, this is a checkpoint

bhalevy replied:
what name would you suggest instead?

denesb replied:
checkpoint

BTW please spell out the return type, don't use `auto`

bhalevy replied:
Done in v2 ([b6705ad](https://github.com/scylladb/scylladb/commit/b6705ad48bc3aebb0a6b8fed495bbc8e30906535))

bhalevy commented:
In v2 ([b6705ad](https://github.com/scylladb/scylladb/commit/b6705ad48bc3aebb0a6b8fed495bbc8e30906535)):
- Moved https://github.com/scylladb/scylla-dtest/issues/5581 to #23153 for backport
- renamed safe_start to checkpoint
- marked as `Fixes #23153`
- rebased

xemul commented on main.cc:
```diff
@@ -537,6 +553,11 @@ void print_starting_message(int ac, char** av, const bpo::parsed_options& opts)
     fmt::print("parsed command line options: {}\n", format_parsed_options(opts.options));
 }
 
+static void checkpoint(stop_signal& stop, sstring what, bool ready = false) {
```
nit: Can be a method of stop_signal itself, thus aligned with stop_signal::ready()

bhalevy replied:
Yeah, this crossed my mind, but then the stop_signal would be involved in supervisor::notify which is out of scope for it.  Ultimately, we may want a mid-level abstraction between main and the stop_signal to encapsulate the checkpoint (and more, as my original work on a service-control unit that will control the services state machine and ependencies)
